core/src/main/java/org/apache/iceberg/actions/BaseSnapshotTable.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.actions;
+
+import org.immutables.value.Value;
+
+@Value.Enclosing
+// https://github.com/immutables/immutables/issues/291 does not apply here because we're not adding
+// any Immutable-specific class to the classpath
+@SuppressWarnings("ImmutablesStyle")
+@Value.Style(typeImmutableEnclosing = "ImmutableSnapshotTable")
+public interface BaseSnapshotTable extends SnapshotTable {
```

Comment:
I forgot to check this in the last PR, but I really don't like that this is a public interface that completely duplicates the `SnapshotTable` interface from API. Is it possible to make this package-private? Or is that a limitation of Immutables?

Reply:
yes that's absolutely possible. Just worth mentioning that the generated class `ImmutableSnapshotTable` will have the same visibility. Did you want to only make this one package-private?

Reply:
Yes, definitely. We want to make sure that people don't see or use these because they aren't real interfaces. They're just artifacts of using Immutables.

Can you also fix the other PR if those were public?

Reply:
it turns out adding `visibility = Value.Style.ImplementationVisibility.PUBLIC` will cause warnings for consuming clients (same issue as #291) and the suggested workaround isn't really feasible for us. 
I've opened https://github.com/immutables/immutables/pull/1474 to fix this in Immutables and verified that consuming clients don't see the below warning with my proposed fix
```
warning: unknown enum constant ImplementationVisibility.PUBLIC
  reason: class file for org.immutables.value.Value$Style$ImplementationVisibility not found
```

Should we go ahead and merge this PR for now to unblock other work and I'll follow up and make those classes package-private but their generated classes public once the issue is fixed upstream?

Reply:
What about just copying the Immutable output classes into src? Then we don't need the duplicate interfaces or codegen not doing what we want.

Reply:
I've looked closer into this and I don't think this is actually a direct issue of using Immutables per-se, but rather something with javac. The generated class looks identical with/without the `ImplementationVisibility` and the javac warning won't cause any issues for consuming clients. 
I went ahead an made all relevant interfaces package-protected and specified the visibility of the generated Class/Builder.